### PR TITLE
feat(api): home za axis first, and home pipettes in parallel

### PR DESCRIPTION
**Pull request by caila-marashaj**
I added a little method in ot3utils containing a list of the axes in the correct order for homing. This could also go in types- is there a preference for location?

### CODE REVIEW

**Review by Laura-Danielle — CHANGES_REQUESTED** [review]
This won't work in the way that you expect for two reasons.

1. move groups are actually dictionaries which _do not_ have guaranteed order
2. You are still sending this all as 1 move group which means that it will try to execute this move group _simultaneously_.

What you need to do is make sure you have at least 2 move groups (1 for the Z/A (pip mounts) and 1 for the rest of the axes).

**Comment by caila-marashaj** [comment]
> 

ohhh ok I understand. On it chief

File: api/src/opentrons/hardware_control/backends/ot3controller.py [hunk 1]
```diff
@@ -220,12 +220,27 @@ async def home(self, axes: List[OT3Axis]) -> OT3AxisMap[float]:
...
         }
+
         speed_settings = (
             self._configuration.motion_settings.max_speed_discontinuity.none
         )
         velocities = {ax: -1 * speed_settings[OT3Axis.to_kind(ax)] for ax in axes}
         group = create_home_group(distances, velocities)
-        runner = MoveGroupRunner(move_groups=[group])
+        move_group = [group]
+        if (OT3Axis.Z_L or OT3Axis.Z_R) in axes:
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/backends/ot3controller.py]: We can use the `OT3Axis.pipette_axes` or `OT3Axis.mount_axes` for this.

File: api/src/opentrons/hardware_control/backends/ot3controller.py [hunk 2]
```diff
@@ -220,12 +220,27 @@ async def home(self, axes: List[OT3Axis]) -> OT3AxisMap[float]:
...
+                ax: distances[ax] for ax in ([OT3Axis.Z_L, OT3Axis.Z_R] and axes)
+            }
+            velocities_z = {
+                ax: velocities[ax] for ax in ([OT3Axis.Z_L, OT3Axis.Z_R] and axes)
+            }
+            for ax in [OT3Axis.Z_L, OT3Axis.Z_R] and axes:
+                del distances[ax]
+                del velocities[ax]
+            group_z = create_home_group(distances_z, velocities_z)
+            move_group.append(group_z)
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/backends/ot3controller.py]: The move group will execute in index order, so the z move would happen after the other one. Also, we're not filtering beforehand so the z axes might still be in the first move

File: api/src/opentrons/hardware_control/backends/ot3controller.py [hunk 3]
```diff
@@ -220,12 +220,27 @@ async def home(self, axes: List[OT3Axis]) -> OT3AxisMap[float]:
         distances = {
             ax: -1 * self.axis_bounds[ax][1] - self.axis_bounds[ax][0] for ax in axes
         }
+
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/backends/ot3controller.py]: better to just create two different distance/velocity dicts right off the bat. We can use the `OT3Axis.pipette_axes` or `OT3Axis.mount_axes` to filter out or grab the Mount axes.

**Review by Laura-Danielle — CHANGES_REQUESTED** [review]
Almost there!

Also, we should add a test for this to make sure that the move runner is receiving the two move groups in the correct order.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
We should home any attached pipettes too. Doesn't really matter when we do it, on the ot2 it's during the Z phase tho

**Comment by sfoster1** [comment]
> We should home any attached pipettes too. Doesn't really matter when we do it, on the ot2 it's during the Z phase tho

You know, unlike the OT2 we have the capability to do far more independent movement commands and a much more easily shareable bus. Because a list of movegroups fed to a single movegrouprunner instance will execute in sequence because the host won't issue an execute of the second until the first have all completed, we could in fact basically fully parallelize the following:

```
sequence 1  |   sequence 2
home z/a    |   home pipettes
then        |
home xy     |
```
by creating _two_ `MoveGroupRunner`s, one for za then xy and the other for pipettes. Make sure that the z/a and x/y homes - unlike all other move groups - do _not_ contain `SingleAxisStep`s for the pipettes. Make sure that the pipettes move group uses a different move group index than za/xy. Then `asyncio.gather` the two `MoveGroupRunner` executions.

You could even _extend_ this parallelization by making sequence 2 also contain the move-to-bottom for the pipettes. This could be really cool!

**Comment by caila-marashaj** [comment]
> > We should home any attached pipettes too. Doesn't really matter when we do it, on the ot2 it's during the Z phase tho
> 
> You know, unlike the OT2 we have the capability to do far more independent movement commands and a much more easily shareable bus. Because a list of movegroups fed to a single movegrouprunner instance will execute in sequence because the host won't issue an execute of the second until the first have all completed, we could in fact basically fully parallelize the following:
> 
> ```
> sequence 1  |   sequence 2
> home z/a    |   home pipettes
> then        |
> home xy     |
> ```
> 
> by creating _two_ `MoveGroupRunner`s, one for za then xy and the other for pipettes. Make sure that the z/a and x/y homes - unlike all other move groups - do _not_ contain `SingleAxisStep`s for the pipettes. Make sure that the pipettes move group uses a different move group index than za/xy. Then `asyncio.gather` the two `MoveGroupRunner` executions.
> 
> You could even _extend_ this parallelization by making sequence 2 also contain the move-to-bottom for the pipettes. This could be really cool!

This does sound like a great opportunity- what do you mean when you say the move-to-bottom? Is that part of the homing routine?

File: api/src/opentrons/hardware_control/backends/ot3controller.py [hunk 4]
```diff
@@ -225,19 +225,56 @@ async def home(self, axes: Sequence[OT3Axis]) -> OT3AxisMap[float]:
...
+            move_group_gantry_z.append(z_move)
+        if distances_gantry and velocities_gantry:
+            gantry_move = create_home_group(distances_gantry, velocities_gantry)
+            move_group_gantry_z.append(gantry_move)
+        if distances_pipette and velocities_pipette:
+            pipette_move = create_home_group(distances_pipette, velocities_pipette)
+            move_group_pipette.append(pipette_move)
+
+        runner_gantry_z = MoveGroupRunner(move_groups=move_group_gantry_z)
+        await runner_gantry_z.run(can_messenger=self._messenger)
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/backends/ot3controller.py]: these will run in sequence. if you instead use [asyncio.gather](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather), they will run concurrently.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Use asyncio.gather instead of running in sequence, and we should test on a robot

**Review by sfoster1 — APPROVED** [review]
Nice! Watched it work on a machine, code and tests look good.

**Review by Laura-Danielle — APPROVED** [review]
looks great to me! Nice tests :)

**Pull request merged**